src/test/java/com/impactupgrade/nucleus/it/MailchimpIT.java

```diff
@@ -0,0 +1,125 @@
+package com.impactupgrade.nucleus.it;
+
+import com.ecwid.maleorang.MailchimpObject;
+import com.ecwid.maleorang.method.v3_0.lists.members.MemberInfo;
+import com.impactupgrade.nucleus.App;
+import com.impactupgrade.nucleus.client.MailchimpClient;
+import com.impactupgrade.nucleus.environment.EnvironmentConfig;
+import com.impactupgrade.nucleus.model.CrmContact;
+import org.apache.commons.lang3.RandomStringUtils;
+import org.junit.jupiter.api.Test;
+
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class MailchimpIT extends AbstractIT {
+
+  protected MailchimpIT() {
+    super(new App(envFactorySfdcMailchimp));
+  }
+
+  @Test
+  public void bulkImportDataToMailchimp() throws Exception {
+    List<String> unsubscribeEmails = List.of("unsub1@gmail.com", "unsub2@gmail.com", "unsub3@gmail.com");
+    List<String> cleanEmails = List.of("clean1@gmail.com", "clean2@gmail.com", "clean3@gmail.com");
+    List<String> emails = new ArrayList<>();
+    emails.addAll(unsubscribeEmails);
+    emails.addAll(cleanEmails);
+
+    // 1. Bulk import to SF
+    List<Object> values = new LinkedList<>();
+    values.add(List.of("Contact First Name", "Contact Last Name", "Contact Email"));
+    emails.forEach(email -> values.add(toCsvList(email)));
+    postToBulkImport(values);
+
+    List<CrmContact> crmContacts = env.primaryCrmService().getContactsByEmails(emails);
+    assertNotNull(crmContacts);
+
+    List<String> crmContactsEmails = crmContacts.stream().map(crmContact -> crmContact.email).collect(Collectors.toList());
+    assertEquals(emails.size(), crmContactsEmails.size());
+    assertTrue(crmContactsEmails.containsAll(emails));
+
+    // 2. Sync to MC
+    EnvironmentConfig.Mailchimp mailchimp = env.getConfig().mailchimp.get(0);
+    String listId = mailchimp.lists.get(0).id;
+    MailchimpClient mailchimpClient = new MailchimpClient(mailchimp, env);
+
+    addEmailsToList(unsubscribeEmails, "subscribed", listId, mailchimpClient);
+    addEmailsToList(cleanEmails, "subscribed", listId, mailchimpClient);
```

Comment:
Rather than using the MC API directly to insert the contacts, what if you first added them to the CRM with their opt-ins checked? Then run the MC sync and ensure those contacts get subscribed to the MC list. After those assertions, then update the contacts to cleaned/unsubscribed, resync, and ensure the appropriate action is taken?

Reply:
Will do